Source/Controls.cs

```diff
@@ -82,31 +85,187 @@ public static void Consume()
         _ => "Xbox Series",
     };
 
-	private static string GetPromptLocation(string name)
-	{
-		var gamepad = Input.Controllers[0].Gamepad;
-
-		if (!prompts.TryGetValue(gamepad, out var list))
-			prompts[gamepad] = list = new();
-
-		if (!list.TryGetValue(name, out var lookup))
-			list[name] = lookup = $"Controls/{GetControllerName(gamepad)}/{name}";
-		
-		return lookup;
-	}
-
-	public static string GetPromptLocation(VirtualButton button)
-	{
-		// TODO: instead, query the button's actual bindings and look up a
-		// texture based on that! no time tho
-		if (button == Confirm)
-			return GetPromptLocation("confirm");
-		else
-			return GetPromptLocation("cancel");
-	}
-
-	public static Subtexture GetPrompt(VirtualButton button)
-	{
-		return Assets.Subtextures.GetValueOrDefault(GetPromptLocation(button));
-	}
-}
\ No newline at end of file
+    private static string GetPromptLocation(string name)
+    {
+        var gamepad = Input.Controllers[0].Gamepad;
+
+        if (!prompts.TryGetValue(gamepad, out var list))
+            prompts[gamepad] = list = new();
+
+        if (!list.TryGetValue(name, out var lookup))
+            list[name] = lookup = $"Controls/{GetControllerName(gamepad)}/{name}";
+
+        return lookup;
+    }
+
+    public static string GetPromptLocation(VirtualButton button)
+    {
+        // TODO: instead, query the button's actual bindings and look up a
+        // texture based on that! no time tho
+        if (button == Confirm)
+            return GetPromptLocation("confirm");
+        else
+            return GetPromptLocation("cancel");
+    }
+
+    public static Subtexture GetPrompt(VirtualButton button)
+    {
+        return Assets.Subtextures.GetValueOrDefault(GetPromptLocation(button));
+    }
+
+
+    private static List<(string, object)> controls = new(8)
+    {
+        ("Move", Move ),
+        ("Menu Move", Menu ),
+        ("Camera", Camera ),
+        ("Jump", Jump ),
+        ("Dash", Dash ),
+        ("Climb", Climb ),
+        ("Menu Confirm", Confirm ),
+        ("Menu Back", Cancel ),
+        ("Pause", Pause )
+    };
+
+#pragma warning disable CS8600
+#pragma warning disable CS8604
+    // i probably should have handled it a bit better but :(, im just hoping that this isn't the slowest code in the game. - Viv
+    public static void HandleCustomControls(JsonNode root)

```

Comment:
Would it not be better to deserialize to a type-checked object?

Reply:
This existing code is extremely volatile and will cause problems if even minor issues are present.